### ISSUE TITLE
dl_unload_all_files() revisited

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1718]
Migrated from [rt.perl.org#2968](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2968) (status was 'open')

Searchable as RT2968$

**Comment by p5pRT** [comment]
### From @AlanBurlison

My current understanding is that this patch broke DBD&#8203;::Oracle on Linux \-
I'm not quite sure why\.  I can see a potential problem if a module uses
call\_atexit\(\)\, and that then uses a bit of XS that has already been
dlclosed\, but for DBD&#8203;::Oracle this doesn't seem to be the case\.

I \(respectfully\!\) disagree with Sarathay's assertion that dlclosing\(\)
stuff when the interpreter exits is an 'unnecessary overhead' \- it is an
\*essential\* part of the cleanup\, much in the same way as freeing up
everything that has been malloced is\.  It is alright to not free and not
dlclose if you know that the process is going to exit anyway\, but in the
case of an embedded perl interpreter this is not necessarily the case\.

I'd like to get the dlclose stuff into a state where it on by default\,
as not having it there generates a particulary abstruse type of bug \- it
took the best part of a year to figure out why mod\_perl dumped core when
built with APXS\.

I'd like some informed suggestions of the best way of doing this
whithout causing collateral damage \- anyone have any ideas?

Alan Burlison

**Comment by p5pRT** [comment]
### From @timbunce

On Sun\, Apr 02\, 2000 at 11&#8203;:05&#8203;:41AM \+0100\, Alan Burlison wrote&#8203;:
> My current understanding is that this patch broke DBD&#8203;::Oracle on Linux \-
> I'm not quite sure why\.

I'll happily help as far as I can if you send me details\.

\(Oracle does very weird things with it's libraries\, and then does
different weird things in the following release\)\.

> I can see a potential problem if a module uses
> call\_atexit\(\)\, and that then uses a bit of XS that has already been
> dlclosed\, but for DBD&#8203;::Oracle this doesn't seem to be the case\.
> 
> I \(respectfully\!\) disagree with Sarathay's assertion that dlclosing\(\)
> stuff when the interpreter exits is an 'unnecessary overhead' \- it is an
> \*essential\* part of the cleanup\, much in the same way as freeing up
> everything that has been malloced is\.  It is alright to not free and not
> dlclose if you know that the process is going to exit anyway\, but in the
> case of an embedded perl interpreter this is not necessarily the case\.

I think Sarathy meant it's an 'unnecessary overhead' when \_not\_ embedding\.

> I'd like to get the dlclose stuff into a state where it on by default\,

When embedding\, yes\.

> as not having it there generates a particulary abstruse type of bug \- it
> took the best part of a year to figure out why mod\_perl dumped core when
> built with APXS\.
> 
> I'd like some informed suggestions of the best way of doing this
> whithout causing collateral damage \- anyone have any ideas?

Do dl\_unload\_all\_files\(\) if Perl\_destruct\_level > 0\.

Tim\.

**Comment by p5pRT** [comment]
### From @Hugmeir

On Sun Apr 02 22&#8203;:51&#8203;:44 2000\, RT\_System wrote&#8203;:
> On Sun\, Apr 02\, 2000 at 11&#8203;:05&#8203;:41AM \+0100\, Alan Burlison wrote&#8203;:
> > My current understanding is that this patch broke DBD&#8203;::Oracle on Linux \-
> > I'm not quite sure why\.
> 
> I'll happily help as far as I can if you send me details\.
> 
> \(Oracle does very weird things with it's libraries\, and then does
> different weird things in the following release\)\.
> 
> > I can see a potential problem if a module uses
> > call\_atexit\(\)\, and that then uses a bit of XS that has already been
> > dlclosed\, but for DBD&#8203;::Oracle this doesn't seem to be the case\.
> > 
> > I \(respectfully\!\) disagree with Sarathay's assertion that dlclosing\(\)
> > stuff when the interpreter exits is an 'unnecessary overhead' \- it is an
> > \*essential\* part of the cleanup\, much in the same way as freeing up
> > everything that has been malloced is\.  It is alright to not free and not
> > dlclose if you know that the process is going to exit anyway\, but in the
> > case of an embedded perl interpreter this is not necessarily the case\.
> 
> I think Sarathy meant it's an 'unnecessary overhead' when \_not\_ embedding\.
> 
> > I'd like to get the dlclose stuff into a state where it on by default\,
> 
> When embedding\, yes\.
> 
> > as not having it there generates a particulary abstruse type of bug \- it
> > took the best part of a year to figure out why mod\_perl dumped core when
> > built with APXS\.
> > 
> > I'd like some informed suggestions of the best way of doing this
> > whithout causing collateral damage \- anyone have any ideas?
> 
> Do dl\_unload\_all\_files\(\) if Perl\_destruct\_level > 0\.
> 
> Tim\.
> 

I can't find the context behind this report\. There\.\. seems to be a
solution for the poster's issue at the end\, but does anyone know if this
is still an issue?

Does anyone know what the issue \_was\_? : /

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri May 04 08&#8203;:52&#8203;:57 2012\, Hugmeir wrote&#8203;:
> On Sun Apr 02 22&#8203;:51&#8203;:44 2000\, RT\_System wrote&#8203;:
> > On Sun\, Apr 02\, 2000 at 11&#8203;:05&#8203;:41AM \+0100\, Alan Burlison wrote&#8203;:
> > > My current understanding is that this patch broke DBD&#8203;::Oracle on
Linux \-
> > > I'm not quite sure why\.
> > 
> > I'll happily help as far as I can if you send me details\.
> > 
> > \(Oracle does very weird things with it's libraries\, and then does
> > different weird things in the following release\)\.
> > 
> > > I can see a potential problem if a module uses
> > > call\_atexit\(\)\, and that then uses a bit of XS that has already been
> > > dlclosed\, but for DBD&#8203;::Oracle this doesn't seem to be the case\.
> > > 
> > > I \(respectfully\!\) disagree with Sarathay's assertion that dlclosing\(\)
> > > stuff when the interpreter exits is an 'unnecessary overhead' \- it
is an
> > > \*essential\* part of the cleanup\, much in the same way as freeing up
> > > everything that has been malloced is\.  It is alright to not free
and not
> > > dlclose if you know that the process is going to exit anyway\, but
in the
> > > case of an embedded perl interpreter this is not necessarily the case\.
> > 
> > I think Sarathy meant it's an 'unnecessary overhead' when \_not\_
embedding\.
> > 
> > > I'd like to get the dlclose stuff into a state where it on by default\,
> > 
> > When embedding\, yes\.
> > 
> > > as not having it there generates a particulary abstruse type of
bug \- it
> > > took the best part of a year to figure out why mod\_perl dumped
core when
> > > built with APXS\.
> > > 
> > > I'd like some informed suggestions of the best way of doing this
> > > whithout causing collateral damage \- anyone have any ideas?
> > 
> > Do dl\_unload\_all\_files\(\) if Perl\_destruct\_level > 0\.
> > 
> > Tim\.
> > 
> 
> I can't find the context behind this report\. There\.\. seems to be a
> solution for the poster's issue at the end\, but does anyone know if this
> is still an issue?
> 
> Does anyone know what the issue \_was\_? : /

From reading the thread\, apparently we \(were|are\) not releasing loaded
shared libraries when we clean up the interpreter\.

This isn't a problem when perl is running standalone\, but is a problem
when perl is embedded\, since we're continuing to use resources after the
interpreter has been deleted\.

As to the proposed solution\, I wonder if freeing the libraries on
interpreter destruction is safe \- it should be sort\-of\-safe where the
underlying library management tools reference count shared libraries
\(true for dlopen and MSWin32\)\, but I don't think it's safe otherwise\.

Also\, there doesn't seem to be an XS entry point \(like boot\_Foo for
module Foo\) to do clean up for a loaded XS module\, which means memory
may leak anyway \(and other worse things\, like dangling signal handlers\.\)

Tony

**Comment by p5pRT** [comment]
### From @nwc10

TL;DR&#8203;: It's a mess\.

On Fri\, Aug 09\, 2013 at 12&#8203;:35&#8203;:41AM \-0700\, Tony Cook via RT wrote&#8203;:
> On Fri May 04 08&#8203;:52&#8203;:57 2012\, Hugmeir wrote&#8203;:
> > On Sun Apr 02 22&#8203;:51&#8203;:44 2000\, RT\_System wrote&#8203;:
> > > On Sun\, Apr 02\, 2000 at 11&#8203;:05&#8203;:41AM \+0100\, Alan Burlison wrote&#8203;:

\[whether at exit the interpreter should attempt to dl\_unload\(\) all files it
dl\_load\(\)ed\]

> > > > I'd like some informed suggestions of the best way of doing this
> > > > whithout causing collateral damage \- anyone have any ideas?
> > > 
> > > Do dl\_unload\_all\_files\(\) if Perl\_destruct\_level > 0\.
> > > 
> > > Tim\.
> > > 
> > 
> > I can't find the context behind this report\. There\.\. seems to be a
> > solution for the poster's issue at the end\, but does anyone know if this
> > is still an issue?

As an aside\, I can't figure out how this has actually come to be recorded
as a ticket in RT\, given that it seems to be a 2 message thread sent just to
the mailing list\, back in 2000&#8203;:

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2000\-04/msg00091\.html

\(Back before the ticketing system was on RT\. So what's the old bug number?\)

> > Does anyone know what the issue \_was\_? : /
> 
> >From reading the thread\, apparently we \(were|are\) not releasing loaded
> shared libraries when we clean up the interpreter\.
> 
> This isn't a problem when perl is running standalone\, but is a problem
> when perl is embedded\, since we're continuing to use resources after the
> interpreter has been deleted\.
> 
> As to the proposed solution\, I wonder if freeing the libraries on
> interpreter destruction is safe \- it should be sort\-of\-safe where the
> underlying library management tools reference count shared libraries
> \(true for dlopen and MSWin32\)\, but I don't think it's safe otherwise\.
> 
> Also\, there doesn't seem to be an XS entry point \(like boot\_Foo for
> module Foo\) to do clean up for a loaded XS module\, which means memory
> may leak anyway \(and other worse things\, like dangling signal handlers\.\)

I've looked at this and the related thread "perl\_atexit\(\) considered useless"

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2000\-09/msg00846\.html


Relevant commits from the period are&#8203;:

commit abb9e9dca5a5f1213886f2e81a42c9a565def727
Author&#8203;: Gurusamy Sarathy \<gsar@&#8203;cpan\.org>
Date&#8203;:   Wed Mar 1 00&#8203;:46&#8203;:44 2000 \+0000

&nbsp;    unload extension shared objects when exiting\, implemented
&nbsp;    only for dl\_dlopen\.xs \(from Alan Burlison\)
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;5381


commit 23d2500b2b45b1beddc8de6ccd7c60068286d061
Author&#8203;: Gurusamy Sarathy \<gsar@&#8203;cpan\.org>
Date&#8203;:   Wed Mar 22 18&#8203;:41&#8203;:50 2000 \+0000

&nbsp;    make unloading of extension shared objects \(change\#5381\) a build
&nbsp;    option \(use "Configure \-Accflags=\-DDL\_UNLOAD\_ALL\_AT\_EXIT" to enable\)
&nbsp;    
&nbsp;    p4raw\-link&#8203;: @&#8203;5381 on //depot/perl&#8203;: abb9e9dca5a5f1213886f2e81a42c9a565def727
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;5885


Both are just before v5\.6\.0 was released\. Nothing has been changed in the code
since\. \(The documentation failed to note the second commit until yesterday\)





It's a mess\.

I can't actually work out what the problem was that Alan was reporting in the
second thread \("perl\_atexit\(\) considered useless"\)\.

&nbsp;    I've been looking at memory errors in perl\, and found that if
&nbsp;    DL\_UNLOAD\_ALL\_AT\_EXIT is defined\, it causes heap problems on exit\.  The
&nbsp;    cause is that the AV holding the references to the dlopen\(\) handles has
&nbsp;    been reclaimed before the perl atexit processing takes place to
&nbsp;    dlclose\(\) the handles\.  I could fix this by keeping a list of the dlopen
&nbsp;    handles locally in DynaLoader\.xs\, but I think there is a more general
&nbsp;    problem here \- what use is perl\_atexit\(\) if it is called after globals
&nbsp;    have been destroyed?


perl\_atexit\(\) is called after object destruction\, but \*before\* all the
unblessed SVs gets destroyed\, and the AV in question is
@&#8203;DynaLoader&#8203;::dl\_librefs\, which is not blessed\. So what he describes shouldn't
happen\.

I've also built blead from that era with gcc 4\.8\.1 and ASAN\.
Yes\, this works\. I was a bit surprised how easy it was\.

The build and most of the tests pass\. There are a couple of bugs tickled by
regression tests which ASAN spots\, but mostly 5\.6\.0 was pretty clean\.
\(No doubt in large part because Alan did a lot of work with purify around
that time to identify and fix problems\.\)
But ASAN can't find any sort of heap error\. So I'm confused\.


I think I can see how the DBD&#8203;::Oracle problem might be happening\. It's
likely the same sort of thing as this bit of code in DynaLoader\.t&#8203;:

&nbsp;    SKIP&#8203;: \{
&nbsp;            skip "unloading unsupported on $^O"\, 2
&nbsp;                if \($old\_darwin || $^O eq 'VMS'\);
&nbsp;            my $module = pop @&#8203;loaded\_modules;
&nbsp;            skip "File&#8203;::Glob sets PL\_opfreehook"\, 2 if $module eq 'File&#8203;::Glob';
&nbsp;            my $r = eval \{ DynaLoader&#8203;::dl\_unload\_file\($libref\) \};

in that\, if DBD&#8203;::Oracle is using something \(probably tie magic\) that is a
pointer into the shared object held by an SV \(or similar\) that survives
until late in global destruction\, then the point at which perl\_atexit\(\) runs
is before then\, and so it creates dangling pointers\.


DL\_UNLOAD\_ALL\_AT\_EXIT is not the default\. I'm not sure if anyone is building
with it\. Certainly\, no\-one is building with it and ithreads\, because there is
an explosion of fail if you do\. The problem is that DynaLoader is using
perl\_atexit\(\) to unmap shared libraries\, and that's per \*interpreter\* exit\.
So as soon as an ithread is spawned\, its exit triggers the unmapping of all
shared objects\, and the next time the parent thread tries to use one\, kaboom\!

I suspect that no\-one is building with it\, because since this \(innocent\)
commit\, 

commit 667763bdbf37a30596512ca0a08a720d86c7e2a8
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Thu Oct 4 21&#8203;:56&#8203;:00 2012 \-0700

&nbsp;    Make PerlIO&#8203;::encoding more resilient to buffer changes


ext/PerlIO\-encoding/t/encoding\.t will start SEGVing\, because something still
in use during global destruction is unmapped\.


The issues are sort\-of threefold

1\) As ithreads are effectively forks as far as the interpreter goes\, but are
&nbsp;   just threads as far as the OS knows\, just calling perl\_atexit\(\) is the
&nbsp;   wrong point to call any sort of unloading\, because it will happen in each
&nbsp;   ithreads
&nbsp;   We \(at least\) need to only do it in the top level interpreter
&nbsp;   However\, we'd still leak if child interpreters dl\_open\(\) more things\, so I
&nbsp;   guess that this means that ideally it ought to be a shared structure
&nbsp;   tracking this stuff
2\) perl\_atexit\(\) fires too early during global destruction to be the right
&nbsp;   place to run this\. As it's C\-only code that needs running\, it ought to be
&nbsp;   really late in global destruction\, after all SVs have been freed\.
&nbsp;   But the point at which perl\_atexit\(\) runs is \(differently\) useful so it
&nbsp;   can't be changed\. It's sort of feeling like we need two \(or more?\) hook
&nbsp;   points in the global destruction sequence
3\) as to dl\_unload\_file\(\) at any other time\, and as demonstrated by
&nbsp;   File&#8203;::Glob\, we have a mess\. Right now\, it's never safe to unload arbitrary
&nbsp;   modules
&nbsp;   a\) because we don't have a way to tell them to free up things
&nbsp;   b\) possibly because they know that they can't \(eg OP free hooks\)
&nbsp;   c\) definitely because the way we currently do hooks\, eg OP free hooks\, we
&nbsp;      encourage modules to store the old hook value\, and restore it later\.
&nbsp;      Which means that any other module might have a pointer to File&#8203;::Glob's
&nbsp;      innards\, which File&#8203;::Glob has no idea about

Point \(3\) doesn't matter for last\-stage global destruction unloading\, which
I think was the specific problem that Alan was describing\, and I suspect is
the least impossible to fix\.



I think to properly solve the "at interpreter exit" cleanup we'd really need
to change dl\_load\_file\(\) to internally track what it's been called for
\(tracking shared across threads\)\, and \(I think\) how often\, with ownership
being upped on thread clone\. At which point\, interpreter exit is safe to call
dl\_unload\_all\_files\(\)\. But only really late\.


The current code is \*in XS\*\, and kicks back to more XS code\. I think that it
could be replaced by pure Perl code \(as\-is\) with no loss of functionality\.
But no actual bug fixes either\.

\#ifdef DL\_UNLOAD\_ALL\_AT\_EXIT
/\* Close all dlopen'd files \*/
static void
dl\_unload\_all\_files\(pTHX\_ void \*unused\)
\{
&nbsp;    CV \*sub;
&nbsp;    AV \*dl\_librefs;
&nbsp;    SV \*dl\_libref;

&nbsp;    if \(\(sub = get\_cvs\("DynaLoader&#8203;::dl\_unload\_file"\, 0\)\) \!= NULL\) \{
&nbsp;        dl\_librefs = get\_av\("DynaLoader&#8203;::dl\_librefs"\, 0\);
&nbsp;        while \(\(dl\_libref = av\_pop\(dl\_librefs\)\) \!= &PL\_sv\_undef\) \{
&nbsp;           dSP;
&nbsp;           ENTER;
&nbsp;           SAVETMPS;
&nbsp;           PUSHMARK\(SP\);
&nbsp;           XPUSHs\(sv\_2mortal\(dl\_libref\)\);
&nbsp;           PUTBACK;
&nbsp;           call\_sv\(\(SV\*\)sub\, G\_DISCARD | G\_NODEBUG\);
&nbsp;           FREETMPS;
&nbsp;           LEAVE;
&nbsp;        \}
&nbsp;    \}
\}
\#endif


Part of the fun of changing things is that it's provable that other code out
there already manipulates @&#8203;Dynaloader&#8203;::dl\_librefs directly&#8203;:

&nbsp;    http&#8203;://grep\.cpan\.me/?q=dl\_librefs

but I'm not sure if all code doing so

1\) only adds \*all\* handles it got by calling dl\_load\_file\(\)
2\) takes no action based on @&#8203;dl\_librefs \- it thinks that it's just doing
&nbsp;   booking for DynaLoader


ie \- I've not looked at the code on CPAN closely enough to work out whether
it would work to change dl\_load\_file\(\) to internally track what it was called
with\, and then have dl\_unload\_all\_files\(\) uses that and ignore
@&#8203;DynaLoader&#8203;::dl\_librefs\.


We possibly also need \*two\* hooks into shared objects\. One for "this
interpreter is closing down"\, which would probably be XS code\, and called
just before object cleanup\. And a second\, C code\, for "the last interpreter
is now about to unload you"\.
Although\, botherit\, we might also want an XS level hook for "we're about to
start global destruction and you will be freed at the end"\.

Problem is that\, like PerlIO\, we want to be able to do things both when there
still is an interpreter \(to be able to use it\)\, and after it's gone \(because
it was relying on things we provide\)


We'd still be trusting shared objects not to leave signal handers live\.
\(and not to use atexit\(\) routines\, and I guess pthread\_atfork\(\) handlers\,
as it looks like neither can be de\-registered\)



To solve the more general problem of dl\_unload\_file\(\) \- ie to be able to
unload shared objects \*before\* global destruction\, I think that we'd also
need to revisit how we're doing hooks \(\*all\* of them\)\, and provide some
mechanism to register handlers for op\_free\, op\_check\, opcode overrides\,
so that \*only\* the interpreter is remembering pointers to extension code\.
We'd then need to "fix" everything on CPAN to be conformant\.

And I think we'd also need those hooks described above\, so that XS modules
can be told "you need to clean up" and undo whatever is necessary to unlink
themselves from the interpreter\.


It's hard work to fix the first problem\.
The "fix everything on CPAN" means that it's impossible to ever completely
fix the second\. However\, having a better hook system would be useful in
its own right\, even if we never get to the point of doing enough to fix
the general dl\_unload\_file\(\) issue\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Aug 09 00&#8203;:35&#8203;:40 2013\, tonyc wrote&#8203;:
> 
> From reading the thread\, apparently we \(were|are\) not releasing loaded
> shared libraries when we clean up the interpreter\.
> 
> This isn't a problem when perl is running standalone\, but is a problem
> when perl is embedded\, since we're continuing to use resources after the
> interpreter has been deleted\.
> 
> As to the proposed solution\, I wonder if freeing the libraries on
> interpreter destruction is safe \- it should be sort\-of\-safe where the
> underlying library management tools reference count shared libraries
> \(true for dlopen and MSWin32\)\, but I don't think it's safe otherwise\.
> 
> Also\, there doesn't seem to be an XS entry point \(like boot\_Foo for
> module Foo\) to do clean up for a loaded XS module\, which means memory
> may leak anyway \(and other worse things\, like dangling signal handlers\.\)
> 

END block? package var with a DESTROY? I think I once did it with an XS
module by having an OS mutex protect a C reference count\, to run the
unloader\, whatever Perl level thing that triggered it has to obtain the
lock\, and the count must be 0\, objects cant initialize from other perl
threads because the mutex is being held\. When the first thread clears
all the static resources and releases the lock\, the 2nd thread will
reinitialize the DLL again\. Also \-1 could be placed in the reference
count to indicate to never reinitialize the DLL again in the process and
return an error/undef on Perl lang level if an object new was tried\. I
am not referring to the Windows DLL loader\, just Perl XS concepts\.

Windows has DllMain which is always called at DLL unload time\, but I
dont think that exists on any other OS\.

> Tony

related link
http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2000/02/msg8051\.html

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, Aug 22\, 2013 at 07&#8203;:09&#8203;:36AM \-0700\, bulk88 via RT wrote&#8203;:
> On Fri Aug 09 00&#8203;:35&#8203;:40 2013\, tonyc wrote&#8203;:
> > 
> > From reading the thread\, apparently we \(were|are\) not releasing loaded
> > shared libraries when we clean up the interpreter\.
> > 
> > This isn't a problem when perl is running standalone\, but is a problem
> > when perl is embedded\, since we're continuing to use resources after the
> > interpreter has been deleted\.
> > 
> > As to the proposed solution\, I wonder if freeing the libraries on
> > interpreter destruction is safe \- it should be sort\-of\-safe where the
> > underlying library management tools reference count shared libraries
> > \(true for dlopen and MSWin32\)\, but I don't think it's safe otherwise\.
> > 
> > Also\, there doesn't seem to be an XS entry point \(like boot\_Foo for
> > module Foo\) to do clean up for a loaded XS module\, which means memory
> > may leak anyway \(and other worse things\, like dangling signal handlers\.\)
> > 
> 
> END block? package var with a DESTROY? I think I once did it with an XS

As Ilya Z noted in one of the threads\, Alan's initial suggestion of END is
way to early\. END is run before objects are cleaned up\. A package var with
a DESTROY would trigger \*while\* objects were being cleared up\, which is still
too early\, as any other objects relying on C or XS code in that library
might not yet be cleared up\. \(This is the same bug as Reini figured out in
Storable\)

Ilya Z noted that it really needs to be done after everything Perl\-like is
cleared up&#8203;:

&nbsp;    Alan Burlison writes&#8203;:
&nbsp;    > > Couldn't it be done in an END block inside DynaLoader?

&nbsp;    Too early\.  The only safe place \(if any\) should be
&nbsp;    after\-the\-end\-of\-global\-destruction \(assuming the list is kept in a C
&nbsp;    structure\, thus is indestructible\)\.

&nbsp;    Ilya

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2000\-01/msg00843\.html


which sort of gets asked again by Sarathy\, to which Alan replies "pass"&#8203;:

&nbsp;    >> There was some doubt in my mind whether call\_atexit\(\) would do the job
&nbsp;    >> properly\.  What happens if the dll has allocated SVs?  The call\_atexit\(\)
&nbsp;    >> callbacks are called after objects are destroyed but \*before\* the SV
&nbsp;    >> arenas are deallocated\.  Do you foresee any problems from that?
&nbsp;    >
&nbsp;    >Pass\.  Does that mean that @&#8203;dl\_librefs will have been reclaimed?
&nbsp;    
&nbsp;    Not unless @&#8203;dl\_librefs is an object\.
&nbsp;    
&nbsp;    >                                                                 If so\,
&nbsp;    >the patch seems to work remarkably well considering it is popping stuff
&nbsp;    >of a free'd AV\.  I don't think the dll allocating SVs has anything to do
&nbsp;    >with it \- dlclose\(\) doesn't do any sort of cleanup \[\.\.\.\]
&nbsp;    
&nbsp;    OK\, I think that answers my question\.
&nbsp;    
&nbsp;    >I wrote the patch based on what I thought was the consensus on the
&nbsp;    >correct way to do this \- was I mislead?
&nbsp;    
&nbsp;    Doesn't look much like it\.  I just want to be sure we aren't overlooking
&nbsp;    any sharp corners\.

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2000\-02/msg01594\.html

I think that we are overlooking potential sharp corners\, and that Ilya Z
remains correct \- if we are going to unload dynamic libraries\, we have to do
it very late\, after we've cleaned up all perl\-related data structures\, just
before we free the interpreter structure\.

> http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2000/02/msg8051\.html

Thanks for the link

Nicholas Clark

**Comment by bulk88** [comment]
https://github.com/Perl/perl5/blob/0eda2a52e6c9ceb462356f5125467832b5fb0e96/ext/DynaLoader/t/DynaLoader.t#L27

https://github.com/Perl/perl5/blob/0eda2a52e6c9ceb462356f5125467832b5fb0e96/ext/DynaLoader/t/DynaLoader.t#L157

Still a problem, adding any more XS deps, to those 6 XS modules, if #7 module, has an XSUB `END()`, `ext/DynaLoader/t/DynaLoader.t` SEGVs.
